src/test/clojure/xtdb/oidc_integration_test.clj

```diff
@@ -0,0 +1,166 @@
+  (:import [org.postgresql.util PSQLException]))
+
+(t/use-fixtures :once
+  (fn [f]
+    (tu/with-container authn-test/container
+      (fn [_]
+        (f)))))
+
+(t/deftest ^:integration test-oidc-password-flow
+  (authn-test/seed! authn-test/container)
+  (let [port (tu/free-port)
+        oidc-config [:openid-connect {:issuer-url (str (.getAuthServerUrl authn-test/container) "/realms/master")
+                                      :client-id "xtdb"
+                                      :client-secret "xtdb-secret"
+                                      :rules [{:user "test-user" :method :password :address "127.0.0.1"}]}]]
+    (with-open [_node (xtn/start-node {:authn oidc-config
+                                       :server {:port port}})]
+      (let [ds-spec {:dbtype "postgresql"
+                     :host "localhost"
+                     :port port
+                     :dbname "xtdb"
+                     :user "test-user"
+                     :password "password124"}]
+
+        (t/testing "successful password authentication"
+          (with-open [conn (jdbc/get-connection ds-spec)]
+            (let [result (jdbc/execute-one! conn ["SELECT 1 as test"])]
+              (t/is (= 1 (:test result))))))
+        
+        (t/testing "failed password authentication"
+          (t/is (thrown-with-msg? PSQLException #"password authentication failed for user: test-user"
+                                  (jdbc/get-connection (assoc ds-spec :password "wrong-password")))))))))
+
+(t/deftest ^:integration test-oidc-client-credentials
+  (let [clients (:clients (authn-test/seed! authn-test/container))
+        {:keys [client-id client-secret]} (:test clients)
+        port (tu/free-port)
+        oidc-config [:openid-connect {:issuer-url (str (.getAuthServerUrl authn-test/container) "/realms/master")
+                                      :client-id "xtdb"
+                                      :client-secret "xtdb-secret"
+                                      :rules [{:user "oid-client" :method :client-credentials}]}]]
+    (with-open [_node (xtn/start-node {:authn oidc-config
+                                       :server {:port port}})]
+      (let [ds-spec {:dbtype "postgresql"
+                     :host "localhost"
+                     :port port
+                     :dbname "xtdb"
+                     :user "oid-client"
+                     :password (authn-test/encode-client-creds client-id client-secret)}]
+
+        (t/testing "successful client credentials authentication"
+          (with-open [conn (jdbc/get-connection ds-spec)]
+            (let [result (jdbc/execute-one! conn ["SELECT 'authenticated' as status"])]
+              (t/is (= "authenticated" (:status result))))))
+
+        
+        (t/testing "failed client credentials authentication with badly encoded credentials"
+          (t/is (thrown-with-msg? PSQLException #"ERROR: client credentials error: Invalid base64 encoding: Illegal base64 character 2d"
```

Comment:
there's a pgw-test util here which unwraps returned anomalies, so you can then test with `(t/is (anomalous? ...))` again